src/Data/Text/Encoding.hs

```diff
@@ -391,7 +391,37 @@ decodeUtf8Lenient = decodeUtf8With lenientDecode
 --
 -- @since 1.1.0.0
 encodeUtf8Builder :: Text -> B.Builder
-encodeUtf8Builder = encodeUtf8BuilderEscaped (BP.liftFixedToBounded BP.word8)
+encodeUtf8Builder =
+    -- manual eta-expansion to ensure inlining works as expected
+    \txt -> B.builder (step txt)
+  where
+    step txt@(Text arr off len) !k br@(B.BufferRange op ope)
+      -- Ensure that the common case is not recursive and therefore yields
+      -- better code.
+      | op' <= ope = do
+          unsafeSTToIO $ A.copyToPointer arr off op len
+          k (B.BufferRange op' ope)
+      | otherwise = textCopyStep txt k br
+      where
+        op' = op `plusPtr` len
+{-# INLINE encodeUtf8Builder #-}
+
+textCopyStep :: Text -> B.BuildStep a -> B.BuildStep a
+textCopyStep (Text arr off0 len) k =
+    go off0
+  where
+    go !off (B.BufferRange op ope)
+      | inpRemaining <= outRemaining = do
+          unsafeSTToIO $ A.copyToPointer arr off op inpRemaining
+          let !br = B.BufferRange (op `plusPtr` inpRemaining) ope
+          k br
+      | otherwise = do
+          unsafeSTToIO $ A.copyToPointer arr off op outRemaining
+          let !off' = off + outRemaining
+          return $ B.bufferFull 1 ope (go off')
+      where
+        outRemaining = ope `minusPtr` op
+        inpRemaining = len - off
```

Comment:
@alexbiehl could you please add a test, which exercises `encodeUtf8Builder` with non-zero `off`?

Reply:
Is there any existing test that does so? No idea how to test that effectively.

Reply:
Ah, got it.

Reply:
And indeed it catched the bug!